### PR TITLE
[enhancement] Make expr and scalar operator convert more easily

### PR DESCRIPTION
**Pull request by kangkaisen**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

1 Currently, when exprs and  scalar operators transform each other, we need  a lot of contexts, which is hard to use, this PR is in order to Make expr and scalar operator convert more easily.
2. Make ExpressionAnalyzer use more easily.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 59 / 94 (62.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SlotDescriptor.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|59|71|83.10%|

File: fe/fe-core/src/main/java/com/starrocks/analysis/SlotDescriptor.java [hunk 1]
```diff
@@ -82,6 +82,14 @@ public SlotDescriptor(SlotId id, TupleDescriptor parent) {
         this.isMultiRef = false;
     }
 
+    public SlotDescriptor(SlotId id, String name, Type type, boolean isNullable) {
+        this.id = id;
+        this.label_ = name;
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/analysis/SlotDescriptor.java:87]: label_ -> label

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java [hunk 2]
```diff
@@ -490,4 +499,17 @@ public ScalarOperator visitCloneExpr(CloneExpr node, Void context) {
...
+    static class IgnoreSlotVisitor extends Visitor {
+        public IgnoreSlotVisitor() {
+            super(new ExpressionMapping(new Scope(RelationId.anonymous(), new RelationFields())),
+                    Collections.EMPTY_LIST);
+        }
+
+        @Override
+        public ScalarOperator visitSlot(SlotRef node, Void context) {
+            return new ColumnRefOperator(node.getSlotId().asInt(),
+                    node.getType(), node.getColumnName(), node.isNullable());
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:512]: IgnoreSlotVisitor seems not handling slot, there is a little mismatch

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java [hunk 3]
```diff
@@ -66,7 +68,15 @@
...
-        return expression.accept(new Formatter(), descTbl);
+        return expression.accept(new Formatter(ScalarOperatorToExpr::buildExecExpression), descTbl);
+    }
+
+    public static Expr buildExprIgnoreSlot(ScalarOperator expression, FormatterContext descTbl) {
+        return expression.accept(new IgnoreSlotFormatter(ScalarOperatorToExpr::buildExprIgnoreSlot), descTbl);
+    }
+
+    interface BuildExpr {
+        Expr build(ScalarOperator expression, FormatterContext descTbl);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java:79]: I didn't find the implement of this interface 👀

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java:79]: new Formatter(ScalarOperatorToExpr::buildExecExpression),

new IgnoreSlotFormatter(ScalarOperatorToExpr::buildExprIgnoreSlot),

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 2 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:512]: IgnoreSlotVisitor ignore the slot, which is consistent with the name

File: fe/fe-core/src/main/java/com/starrocks/analysis/SlotDescriptor.java (hunk 1 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/analysis/SlotDescriptor.java:87]: The old codes are all label_, This PR won't change it, we could change it in another PR

**Comment by kangkaisen** [comment]
> I recommend extracting the logic of mutual conversion separately, rather than mixing it in ExpressionAnalyzer

The logic is very clear, no any mixing.

**Comment by kangkaisen** [comment]
run starrocks_be_unittest

**Comment by kangkaisen** [comment]
run starrocks_clang-format

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java [hunk 4]
```diff
@@ -66,7 +68,15 @@
...
     public static Expr buildExecExpression(ScalarOperator expression, FormatterContext descTbl) {
-        return expression.accept(new Formatter(), descTbl);
+        return expression.accept(new Formatter(ScalarOperatorToExpr::buildExecExpression), descTbl);
+    }
+
+    public static Expr buildExprIgnoreSlot(ScalarOperator expression, FormatterContext descTbl) {
+        return expression.accept(new IgnoreSlotFormatter(ScalarOperatorToExpr::buildExprIgnoreSlot), descTbl);
+    }
+
+    interface BuildExpr {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java:78]:
```suggestion
    @FunctionalInterface
    interface BuildExpr {
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 59 / 94 (62.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SlotDescriptor.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|59|71|83.10%|

**Review by Youngwb — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**